docs/CodeTransformation.md

```diff
@@ -0,0 +1,102 @@
+---
+id: code-transformation
+title: Code Transformation
+---
+
+Jest runs the code in your project as JavaScript, but if you use some syntax not supported by Node.js out of the box (such as JSX, types from TypeScript, Vue templates etc.) then you'll need to transform that code into plain JavaScript, similar to what you would do when building for browsers.
+
+Jest supports this via the [`transform` configuration option](Configuration.md#transform-objectstring-pathtotransformer--pathtotransformer-object).
+
+A transformer is a module that provides a synchronous function for transforming source files. For example, if you wanted to be able to use a new language feature in your modules or tests that aren't yet supported by Node, you might plug in one of many compilers that compile a future version of JavaScript to a current one.
+
+Jest will cache the result of a transformation and attempt to invalidate that result based on a number of factors, such as the source of the file being transformed and changing configuration.
+
+## Defaults
+
+Jest ships with one transformer out of the box - `babel-jest`. It will automatically load your project's Babel configuration and transform any file matching the following RegEx: `/\.[jt]sx?$/` meaning any `.js`, `.jsx`, `.ts` and `.tsx` file. In addition, `babel-jest` will inject the Babel plugin necessary for mock hoisting talked about in [ES Module mocking](ManualMocks.md#using-with-es-module-imports).
+
+If you override the `transform` configuration option `babel-jest` will no longer be active, and you'll need to add it manually if you wish to use Babel.
+
+## Writing custom transformers
+
+You can write you own transformer. The API of a transformer is as follows:
+
+```ts
+interface Transformer {
+  /**
+   * Indicates if the transformer is capabale of instrumenting the code for code coverage.
+   *
+   * If V8 coverage is _not_ active, and this is `true`, Jest will assume the code is instrumented.
+   * If V8 coverage is _not_ active, and this is `false`. Jest will instrument the code returned by this transformer using Babel.
+   */
+  canInstrument?: boolean;
+  createTransformer?: (options?: unknown) => Transformer;
```

Comment:
Currently options has type `TransformOptions | null` in `Babel-jest`. Should this type be like that ?

Reply:
Not really, the real type is `unknown` - the user can pass whatever they want. We can make it generic, though

Reply:
Ye I think better put generic with default is `unknown`

Reply:
meh, turning it generic turned out to be somewhat painful

Reply:
[`e747197` (#10834)](https://github.com/facebook/jest/pull/10834/commits/e747197a328b6ef986e25cf7c3edad04fa6eca3d)